src/disco/topo/fd_topob.c

```diff
@@ -392,6 +393,95 @@ validate( fd_topo_t const * topo ) {
   }
 }
 
+void
+fd_topob_auto_layout( fd_topo_t * topo ) {
+  /* Incredibly simple automatic layout system for now ... just assign
+     tiles to CPU cores in NUMA sequential order, except for a few tiles
+     which should be floating. */
+
+  char const * FLOATING[] = {
+    "metric",
+    "bencho",
+    "bhole",  // FIREDANCER only
+  };
+
+  char const * ORDERED[] = {
+    "benchg",
+    "benchs",
+    "net",
+    "quic",
+    "verify",
+    "dedup",
+    "pack",
+    "bank",
+    "poh",
+    "pohi",   // FIREDANCER only
+    "shred",
+    "store",
+    "storei", // FIREDANCER only
+    "sign",
+    "gossip", // FIREDANCER only
+    "repair", // FIREDANCER only
+    "replay", // FIREDANCER only
+    "thread", // FIREDANCER only
+    "sender", // FIREDANCER only
+  };
+
+  for( ulong i=0UL; i<topo->tile_cnt; i++ ) {
+    fd_topo_tile_t * tile = &topo->tiles[ i ];
+    tile->cpu_idx = ULONG_MAX;
+  }
+
+  ulong cpu_ordering[ USHORT_MAX ] = { 0UL };
+  ulong num_cpus = fd_numa_cpu_cnt();
+
+  ulong next_cpu_idx = 0UL;
+  for( ulong i=0UL; i<fd_numa_node_cnt(); i++ ) {
+    for( ulong j=0UL; j<num_cpus; j++ ) {
+      ulong numa_node = fd_numa_node_idx( j );
+      if( FD_UNLIKELY( numa_node!=i ) ) continue;
+      cpu_ordering[ next_cpu_idx++ ] = j;
+    }
+  }
+
+  FD_TEST( next_cpu_idx==num_cpus );
+
+  ulong cpu_idx = 0UL;
+  for( ulong i=0UL; i<sizeof(ORDERED)/sizeof(ORDERED[0]); i++ ) {
+    for( ulong j=0UL; j<topo->tile_cnt; j++ ) {
+      fd_topo_tile_t * tile = &topo->tiles[ j ];
+      if( !strcmp( tile->name, ORDERED[ i ] ) ) {
+        if( FD_UNLIKELY( cpu_idx>=num_cpus ) ) {
+          FD_LOG_WARNING(( "auto layout cannot affine tile `%s:%lu` because all the CPUs are already assigned. Leaving it floating", tile->name, tile->kind_id ));
```

Comment:
```suggestion
          FD_LOG_WARNING(( "auto layout cannot set affinity for tile `%s:%lu` because all the CPUs are already assigned. Leaving it floating", tile->name, tile->kind_id ));
```
I remember you like using affine as a verb, but I checked the dictionary, and I don't see that as a real word 🤣

Reply:
Are you sure we want this behavior though? If all the cores have processes pinned with high priority, won't the floating tiles get very little CPU time? In the non-auto case, we just error out and print a message that they need more cores.